internal/app/jwt.go

```diff
@@ -2,223 +2,74 @@ package app
 
 import (
 	"context"
-	"encoding/json"
-	"errors"
-	"net/http"
-	"strings"
-	"sync"
-
-	jwtmiddleware "github.com/auth0/go-jwt-middleware"
-	// TODO: github.com/form3tech-oss/jwt-go is decrepated.
-	// Alternative is https://github.com/golang-jwt/jwt, but go-jwt-middleware still uses github.comform3tech-oss/jwt-go
-	// See also https://github.com/auth0/go-jwt-middleware/issues/73
-	"github.com/form3tech-oss/jwt-go"
+	"net/url"
+	"time"
+
+	jwtmiddleware "github.com/auth0/go-jwt-middleware/v2"
+	"github.com/auth0/go-jwt-middleware/v2/jwks"
+	"github.com/auth0/go-jwt-middleware/v2/validator"
 	"github.com/labstack/echo/v4"
 	"github.com/reearth/reearth-backend/pkg/log"
 )
 
 type contextKey string
 
 const (
-	userProfileKey                     = "auth0_user"
-	debugUserHeader                    = "X-Reearth-Debug-User"
-	contextAuth0AccessToken contextKey = "auth0AccessToken"
-	contextAuth0Sub         contextKey = "auth0Sub"
-	contextUser             contextKey = "reearth_user"
+	debugUserHeader            = "X-Reearth-Debug-User"
+	contextAuth0Sub contextKey = "auth0Sub"
+	contextUser     contextKey = "reearth_user"
 )
 
-type JSONWebKeys struct {
-	Kty string   `json:"kty"`
-	Kid string   `json:"kid"`
-	Use string   `json:"use"`
-	N   string   `json:"n"`
-	E   string   `json:"e"`
-	X5c []string `json:"x5c"`
-}
-
-type Jwks interface {
-	GetJwks(string) ([]JSONWebKeys, error)
-}
-
-type JwksSyncOnce struct {
-	jwks []JSONWebKeys
-	once sync.Once
-}
-
-func (jso *JwksSyncOnce) GetJwks(publicKeyURL string) ([]JSONWebKeys, error) {
-	var err error
-	jso.once.Do(func() {
-		jso.jwks, err = fetchJwks(publicKeyURL)
-	})
+// Validate the access token and inject the user clams into ctx
+func jwtEchoMiddleware(cfg *ServerConfig) echo.MiddlewareFunc {
 
+	issuerURL, err := url.Parse("https://" + cfg.Config.Auth0.Domain)
 	if err != nil {
-		return nil, err
+		log.Fatalf("failed to parse the issuer url: %v", err)
 	}
 
-	return jso.jwks, nil
-}
-
-func fetchJwks(publicKeyURL string) ([]JSONWebKeys, error) {
-	resp, err := http.Get(publicKeyURL)
-	var res struct {
-		Jwks []JSONWebKeys `json:"keys"`
-	}
+	provider := jwks.NewCachingProvider(issuerURL, 60*time.Minute)
 
+	// Set up the validator.
+	jwtValidator, err := validator.New(
+		provider.KeyFunc,
+		validator.RS256,
+		cfg.Config.AuthClient.ISS[0],
```

Comment:
This implementation does not support multiple isses; either change the PR title to "Update go-jwt-middleware to v2" and complete the PR once with just the update, or implement multiple iss support. You will probably have to write your own validator.